### PR TITLE
hashindex: Use Python I/O

### LINKED ISSUE #1809: hashindex size limitations
The size of it is limited to 2**31 entries, probably not important right now (since that would need at least ~180 GB of usable memory). [hash_sizes](https://github.com/borgbackup/borg/blob/master/src/borg/_hashindex.c#L49)

We use normal C I/O currently, which will be limited to 32 bit offsets depending on platform (I think Linux 32 bits and Windows?). This would limit the repository index to ~60 million entries and the chunks cache to ~49 million entries. This is already addressed by the patches in #1688, since Python I/O always uses large offsets if available as far as I'm aware.

### PR DESCRIPTION
**Pull request by enkore**
- Preparation for #1688 / #1101
- Support hash indices >2 GB (fixes #1809)
- Better error reporting

---

Split off from #1688 for easier integration.

The downside is that hashindex.c now includes Python, so one needs to supply the appropriate -lpythonmX.Y etc. to compile it for stand-alone testing. The code is also a bit more complex, because it's (supposed to) correctly clean up references to objects on errors.

### CODE REVIEW

File: src/borg/repository.py [hunk 1]
```diff
@@ -324,9 +324,8 @@ def open_index(self, transaction_id, auto_recover=True):
         index_path = os.path.join(self.path, 'index.%d' % transaction_id).encode('utf-8')
         try:
             return NSIndex.read(index_path)
-        except RuntimeError as error:
-            assert str(error) == 'hashindex_read failed'  # everything else means we're in *deep* trouble
-            logger.warning('Repository index missing or corrupted, trying to recover')
+        except (ValueError, OSError) as exc:
```
Comment (enkore) [inline src/borg/repository.py]: The diff also suggests that the previous error handling was incomplete, see above:

```py
-                if errno:
-                    PyErr_SetFromErrnoWithFilename(OSError, path)
```

Which was not handled here.

File: src/borg/_hashindex.c [hunk 2]
```diff
@@ -248,99 +262,130 @@ count_empty(HashIndex *index)
...
-        }
-        else {
-            EPRINTF_MSG_PATH(path, "fread header failed (expected %ju, got %ju)",
-                             (uintmax_t) sizeof(HashHeader), (uintmax_t) bytes_read);
-        }
-        goto fail;
+    PyObject *header_bytes = PyObject_CallMethod(file, "read", "n", sizeof(HashHeader));
+    if(!header_bytes) {
+        /* Python exception occurred; just return */
+        return 0;
```
Comment (ThomasWaldmann) [inline src/borg/_hashindex.c]: maybe there could be a "goto fail" jumping to the "return index" (which is NULL then)"?

File: src/borg/_hashindex.c [hunk 3]
```diff
@@ -248,99 +262,130 @@ count_empty(HashIndex *index)
...
+        /* TypeError, not a bytes() object */
+        goto fail_decref_header;
     }
-    if((length = ftell(fd)) < 0) {
-        EPRINTF_PATH(path, "ftell failed");
-        goto fail;
+    if(bytes_read != sizeof(HashHeader)) {
+        /* Truncated file */
+        /* Note: %zd is the format for Py_ssize_t */
+        PyErr_Format(PyExc_ValueError, "Could not read header (expected %zd, but read %zd bytes)", sizeof(HashHeader), bytes_read);
```
Comment (ThomasWaldmann) [inline src/borg/_hashindex.c]: sizeof() result also works with %zd?

File: src/borg/_hashindex.c [hunk 4]
```diff
@@ -248,99 +262,130 @@ count_empty(HashIndex *index)
...
+        goto fail_decref_header;
     }
-    if(fseek(fd, sizeof(HashHeader), SEEK_SET) < 0) {
-        EPRINTF_PATH(path, "fseek failed");
-        goto fail;
+
+    /* Find length of file */
+    PyObject *length_object = PyObject_CallMethod(file, "seek", "ni", 0, SEEK_END);
```
Comment (ThomasWaldmann) [inline src/borg/_hashindex.c]: maybe it would be a nice convention for *.c to always have _obj suffix if it is a python object.

File: src/borg/_hashindex.c [hunk 5]
```diff
@@ -248,99 +262,130 @@ count_empty(HashIndex *index)
...
+     * we have them backed by a Python bytes() object instead, and go through
+     * Python I/O.
+     *
+     * Note: Issuing read(buckets_length) is okay here, because buffered readers
+     * will issue multiple underlying reads if necessary. This supports indices
+     * >2 GB on Linux. We also compare lengths later.
+     */
+    PyObject *bucket_bytes = PyObject_CallMethod(file, "read", "n", buckets_length);
+    if(!bucket_bytes) {
+        /* Python exception occured; clean up and return */
```
Comment (ThomasWaldmann) [inline src/borg/_hashindex.c]: occurred

File: src/borg/_hashindex.c [hunk 6]
```diff
@@ -248,99 +262,130 @@ count_empty(HashIndex *index)
...
+
+    index->num_entries = _le32toh(header->num_entries);
+    index->num_buckets = _le32toh(header->num_buckets);
+    index->key_size = header->key_size;
+    index->value_size = header->value_size;
     index->bucket_size = index->key_size + index->value_size;
     index->lower_limit = get_lower_limit(index->num_buckets);
     index->upper_limit = get_upper_limit(index->num_buckets);
-    index->min_empty = get_min_empty(index->num_buckets);
-    index->num_empty = count_empty(index);
```
Comment (ThomasWaldmann) [inline src/borg/_hashindex.c]: why did you delete these 2 lines? in 370 the values are used.

File: src/borg/_hashindex.c [hunk 7]
```diff
@@ -248,99 +262,130 @@ count_empty(HashIndex *index)
...
+    Py_DECREF(bucket_bytes);
+fail_release_header_buffer:
+    PyBuffer_Release(&header_buffer);
+fail_free_index:
+    if(PyErr_Occurred()) {
+        free(index);
     }
+fail_decref_header:
+    Py_DECREF(header_bytes);
     return index;
```
Comment (ThomasWaldmann) [inline src/borg/_hashindex.c]: "fail:" label above this line?

File: src/borg/_hashindex.c [hunk 8]
```diff
@@ -349,7 +352,7 @@ hashindex_read(PyObject *file)
      */
     PyObject *bucket_bytes = PyObject_CallMethod(file, "read", "n", buckets_length);
     if(!bucket_bytes) {
-        /* Python exception occured; clean up and return */
+        /* Python exception occurred; clean up and return */
```
Comment (ThomasWaldmann) [inline src/borg/_hashindex.c]: another one in 382

File: src/borg/_hashindex.c [hunk 9]
```diff
@@ -248,99 +262,144 @@ count_empty(HashIndex *index)
...
     index->num_empty = count_empty(index);
+
     if(index->num_empty < index->min_empty) {
         /* too many tombstones here / not enough empty buckets, do a same-size rebuild */
         if(!hashindex_resize(index, index->num_buckets)) {
-            free(index->buckets);
-            free(index);
-            index = NULL;
-            goto fail;
+            PyErr_Format(PyExc_ValueError, "Failed to rebuild table");
```
Comment (enkore) [inline src/borg/_hashindex.c]: Good thinking

**Comment by ThomasWaldmann** [comment]
Removing the backport label, see #2498.
